core/src/main/java/org/elasticsearch/common/xcontent/support/AbstractXContentParser.java

```diff
@@ -176,14 +194,31 @@ public long longValue(boolean coerce) throws IOException {
             try {
                 return Long.parseLong(stringValue);
             } catch (NumberFormatException e) {
-                return (long) Double.parseDouble(stringValue);
+                return preciseLongValue(stringValue, coerce);
             }
         }
         long result = doLongValue();
         ensureNumberConversion(coerce, result, Long.class);
         return result;
     }
 
+    public static long preciseLongValue(String stringValue, boolean coerce) {
```

Comment:
if you want to use it from different classes, please move it to the `o.e.common.Numbers` class